File: FRAudioUnit/FRAudioUnit/FRAudioUnit.m
    //
//  FRAudioUnit.m
//  TestCase_AudioController
//
//  Created by 상우 강 on 2016. 5. 27..
//  Copyright © 2016년 상우 강. All rights reserved.
//

#import "FRAudioUnit.h"

@implementation FRAudioUnit

-(id)initWithType:(kFRAudioUnitType)type {
    if (self = [super init]) {
        OSStatus err = noErr;
        
        _type = type;
        
        AudioComponentDescription desc;
        desc.componentManufacturer = kAudioUnitManufacturer_Apple;
        desc.componentFlags = 0;
        desc.componentFlagsMask = 0;
        
        switch (_type) {
            case kFRAudioUnitType_Output:
                desc.componentType = kAudioUnitType_Output;
                desc.componentSubType = kAudioUnitSubType_DefaultOutput;
                break;
                
            case kFRAudioUnitType_Mic:
                desc.componentType = kAudioUnitType_Output;
                desc.componentSubType = 'rioc';
                break;
                
            case kFRAudioUnitType_Mixer:
                desc.componentType = kAudioUnitType_Mixer;
                desc.componentSubType = kAudioUnitSubType_MultiChannelMixer;
                break;
                
            case kFRAudioUnitType_Sampler:
                desc.componentType = kAudioUnitType_MusicDevice;
                desc.componentSubType = kAudioUnitSubType_Sampler;
                break;
                
            case kFRAudioUnitType_EQ:
                desc.componentType = kAudioUnitType_Effect;
                desc.componentSubType = kAudioUnitSubType_NBandEQ;
                break;
                
            case kFRAudioUnitType_Reverb:
                desc.componentType = kAudioUnitType_Effect;
                desc.componentSubType = 'rvb2';
                break;
                
            case kFRAudioUnitType_File:
                desc.componentType = kAudioUnitType_Generator;
                desc.componentSubType = kAudioUnitSubType_AudioFilePlayer;
                break;
                
            case kFRAudioUnitType_PeakLimiter:
                desc.componentType = kAudioUnitType_Effect;
                desc.componentSubType = kAudioUnitSubType_PeakLimiter;
                break;
                
            case kFRAudioUnitType_Compressor:
                desc.componentType = kAudioUnitType_Effect;
                desc.componentSubType = kAudioUnitSubType_DynamicsProcessor;
                break;
                
            case kFRAudioUnitType_Converter:
                desc.componentType = kAudioUnitType_FormatConverter;
                desc.componentSubType = kAudioUnitSubType_AUConverter;
                
            default:
                break;
        }
        
        AudioComponent comp = AudioComponentFindNext(NULL, &desc);
        err = AudioComponentInstanceNew(comp, &_audioUnit);
        if (err)
            NSLog(@"New Audio component instance failed : %d", (int)err);
        
        _isInit = NO;
    }
    
    return self;
}

-(void)connectAudioUnit:(FRAudioUnit *)unit withDestinationNumber:(UInt32)destinationNumber {
    OSStatus err = noErr;
    
    AudioUnit sourceUnit = _audioUnit;
    AudioUnit destUnit = [unit audioUnit];
    
    AudioUnitConnection connection = {};
    connection.sourceAudioUnit = sourceUnit;
    if (_type == kFRAudioUnitType_Mic)
        connection.sourceOutputNumber = 1;
    else
        connection.sourceOutputNumber = 0;
    connection.destInputNumber = destinationNumber;
    
    err = AudioUnitSetProperty(destUnit, kAudioUnitProperty_MakeConnection, kAudioUnitScope_Input, 0, &connection, sizeof(connection));
    if (err)
        NSLog(@"make audiounit connection failed : %d", (int)err);
}

-(void)initializeAudioUnit {
    if (_isInit)
        return;
    
    OSStatus err = noErr;
    AudioUnit audioUnit = [self audioUnit];
    err = AudioUnitInitialize(audioUnit);
    if (err)
        NSLog(@"Audio Unit initialized failed : %d", (int)err);
    else
        _isInit = YES;
}

-(void)uninitialize {
    OSStatus err = noErr;
    err = AudioUnitUninitialize(self.audioUnit);
    if (err)
        NSLog(@"audioUnit un initialize failed : %d", (int)err);
    else
        _isInit = NO;
}


@end

File: FRAudioUnit/FRAudioUnit/lib/FRAudioUnit.swift
//
//  FRAudioUnit.swift
//  FRAudioUnit
//
//  Created by exs-mobile 강상우 on 02/08/2019.
//  Copyright © 2019 강상우. All rights reserved.
//

import Foundation
import AVFoundation
import AVKit
import AudioToolbox

enum FRAudioUnitType: Int {
    case IO
    case Mixer
    case Sampler
    case Reverb
    case EQ
    case File
    case PeakLimiter
    case Compressor
    case Converter
}

class FRAudioUnit: NSObject {
    public typealias FourCharCode = UInt32
    public typealias OSType = FourCharCode
    
    var type                    : FRAudioUnitType!
    var audioUnit               : AudioUnit?        = nil
    
    //MARK: - Initializer
    /**
     초기화 함수
     */
    init(_ type: FRAudioUnitType!) {
        super.init()
        self.type = type
        
        var desc = self.getDesc(type)
        let comp = AudioComponentFindNext(nil, &desc)
        let err = AudioComponentInstanceNew(comp!, &audioUnit)
        printError("audio unit create error", err)
    }
    
    //MARK: - Module Private
    func printError(_ title: String, _ err: Any) {
        print("FRAudioUnit ## ", err)
    }
    
    //MARK: - Private
    private func getDesc(_ type: FRAudioUnitType) -> AudioComponentDescription {
        let manufacturer: OSType    = kAudioUnitManufacturer_Apple
        let flags: UInt32           = 0
        let flagsMask: UInt32       = 0
        var componentType: OSType
        var componentSubType: OSType
        
        switch type {
        case .IO:
            componentType       = kAudioUnitType_Output
            componentSubType    = kAudioUnitSubType_RemoteIO
            
        case .Mixer:
            componentType       = kAudioUnitType_Mixer
            componentSubType    = kAudioUnitSubType_MultiChannelMixer
            
        case .Sampler:
            componentType       = kAudioUnitType_MusicDevice
            componentSubType    = kAudioUnitSubType_Sampler
            
        case .EQ:
            componentType       = kAudioUnitType_Effect
            componentSubType    = kAudioUnitSubType_NBandEQ
            
        case .Reverb:
            componentType       = kAudioUnitType_Effect
            componentSubType    = kAudioUnitSubType_Reverb2
            
        case .File:
            componentType       = kAudioUnitType_Output
            componentSubType    = kAudioUnitSubType_VoiceProcessingIO
            
        case .PeakLimiter:
            componentType       = kAudioUnitType_Output
            componentSubType    = kAudioUnitSubType_VoiceProcessingIO
            
        case .Compressor:
            componentType       = kAudioUnitType_Output
            componentSubType    = kAudioUnitSubType_VoiceProcessingIO
            
        case .Converter:
            componentType       = kAudioUnitType_Output
            componentSubType    = kAudioUnitSubType_VoiceProcessingIO
        }
        return AudioComponentDescription(componentType: componentType, componentSubType: componentSubType, componentManufacturer: manufacturer, componentFlags: flags, componentFlagsMask: flagsMask)
    }
    
    //MARK: - Public
    /**
    */
    public func connectAudioUnit(_ destination: FRAudioUnit, _ destInputNumber: UInt32) {
        let source = self.audioUnit
        let dest = destination.audioUnit!
        var sourceOutputNumber: UInt32 = 0
        if self.type == .IO {
            sourceOutputNumber = 1
        }
        var connectionInfo = AudioUnitConnection(sourceAudioUnit: source, sourceOutputNumber: sourceOutputNumber, destInputNumber: destInputNumber)
        
        let size = UInt32(MemoryLayout.size(ofValue: connectionInfo))
        let err = AudioUnitSetProperty(dest, kAudioUnitProperty_MakeConnection, kAudioUnitScope_Input, 0, &connectionInfo, size)
        printError("connect Audio Unit", err)
    }
    
    public func initializeAudioUnit() {
        var err: OSStatus = noErr
        if self.audioUnit == nil {
            printError("initialize audio unit", "unit not allocated")
        }
        err = AudioUnitInitialize(self.audioUnit!)
        printError("initialize audio unit", err)
    }
    
    public func unInitializeAudioUnit() {
        var err: OSStatus = noErr
        err = AudioUnitUninitialize(self.audioUnit!)
        printError("unInitialize audio unit", err)
    }
}
